lib/GenomeFileUtil/core/GenomeInterface.py

```diff
@@ -32,18 +37,78 @@ def __init__(self, config):
         self.scratch = config.raw['scratch']
         self.ws_large_data = WsLargeDataIO(self.callback_url)
 
-    @staticmethod
-    def _validate_save_one_genome_params(params):
+    def save_one_genome(self, params):
+        mass_params = self._set_up_single_params(params)
+        return self._save_genome_mass(mass_params)[0]
+
+    def save_genome_mass(self, params):
+        self._validate_mass_params(params)
+        return self._save_genome_mass(params)
+
+    def _set_up_single_params(self, params):
+        inputs = dict(params)
+        self._validate_genome_input_params(inputs)
+        ws_id = get_int(inputs.pop(_WSID, None), _WSID)
+        ws_name = inputs.pop(_WS, None)
+        if bool(ws_id) == bool(ws_name):  # xnor
+            raise ValueError(f"Exactly one of a '{_WSID}' or a '{_WS}' parameter must be provided")
+        if not ws_id:
+            print(f"Translating workspace name {ws_name} to a workspace ID. Prefer submitting "
+                  + "a workspace ID over a mutable workspace name that may cause race conditions")
+            ws_id = self.dfu.ws_name_to_id(ws_name)
+        mass_params = {_WSID: ws_id, _INPUTS: [inputs]}
+        return mass_params
+
+    def _validate_mass_params(self, params):
```

Comment:
The _setup_single_params and _validate_mass_params look very familiar to me. Are there similar methods in the codebase somewhere?

Reply:
inside GenbankToGenome.py file

Reply:
Can they be shared to avoid duplicating code?

Reply:
You mean moving them into MiscUtils.py and passing additional arguments such as ws_name_to_id and validate_params methods?

Reply:
Something like that, not necessarily in misc utils but in a shard location. Not sure if that's a reasonable thing to do or not, depends how similar the code is - if it's identical then it's easy.

Why would you need to pass in methods?

Reply:
Because they use different validation methods inside.

Reply:
how different are they? can you abstract the two call trees into a single call tree?

Reply:
👍